### PR TITLE
Adjust version commit detection logic

### LINKED ISSUE #58: seems to be choosing "prior" (wrong) commit to git tag for a release
Testing on 000003 (was pushed to https://github.com/dandisets/000003 a few hours back).

When I ran 

```
flock -E 0 -e -n /home/dandi/.run/backup2datalad-cron.lock bash -c '/mnt/backup/dandi/dandisets/tools/backups2datalad-cron 000003'
```

on drogon, it produced a tag, which was not pushed (should have been I guess),  which I now pushed (we can delete later) which points to prior commit
```
dandi@drogon:/mnt/backup/dandi/dandisets/000003$ git show 0.210812.1448 | head 
tag 0.210812.1448
Tagger: DANDI Meta-user <dandi@mit.edu>
Date:   Thu Aug 12 14:48:34 2021 -0400

Version 0.210812.1448 of Dandiset 000003

commit 30680bd1657172824c64e83eac8f02cb33307897
Author: DANDI Meta-user <dandi@mit.edu>
Date:   Tue Aug 10 03:16:35 2021 -0400
```
where metadata didn't have `publishedBy` for assets yet... 

Even after I removed tag and reran -- consistently was tagging that commit instead of the last one which either should be the one, or an off-master commit with the "state" of that particular release being there.

<details>
<summary>I have not done full comparison of the state of assets, but first page is identical ;)
</summary> 

```shell
$> curl --silent -X GET "https://api.dandiarchive.org/api/dandisets/000003/versions/0.210812.1448/assets/" -H "accept: application/json" | jq . > /tmp/published.json
1 13330.....................................:Thu 12 Aug 2021 02:58:22 PM EDT:.
lena:/tmp
$> curl --silent -X GET "https://api.dandiarchive.org/api/dandisets/000003/versions/draft/assets/" -H "accept: application/json" | jq . > /tmp/draft.json  
1 13331.....................................:Thu 12 Aug 2021 02:58:34 PM EDT:.
lena:/tmp
$> md5sum /tmp/draft.json /tmp/published.json
e78164fb26c166fe7711867fd7b8483c  /tmp/draft.json
f5b964485a06f5c09adc8295167b1e18  /tmp/published.json
1 13332.....................................:Thu 12 Aug 2021 02:58:39 PM EDT:.
lena:/tmp
$> diff /tmp/draft.json /tmp/published.json  
3c3
<   "next": "https://api.dandiarchive.org/api/dandisets/000003/versions/draft/assets/?page=2",
---
>   "next": "https://api.dandiarchive.org/api/dandisets/000003/versions/0.210812.1448/assets/?page=2",

```
</details>

and clearly that prior commit didn't have matching listing of assets since back then there were no publishedBy.

PS I will disable autodataladification until this is addressed

### PR DESCRIPTION
**Pull request by jwodder**
Closes #58.

### CODE REVIEW

**Comment by yarikoptic** [comment]
Did you try it out on a few dandisets (without pushing)

**Comment by jwodder** [comment]
@yarikoptic I only tried it out on 000003, and it worked properly.

**Comment by yarikoptic** [comment]
please try on a few others (we have a number of them published already) so we don't get surprises

**Comment by jwodder** [comment]
@yarikoptic I tried it out on 000029 and 000121 as well, and it seemed to work.

File: tools/backups2datalad.py [hunk 1]
```diff
@@ -462,52 +451,81 @@ def mkrelease(
...
+            if (
+                cmt := readcmd(
+                    "git",
+                    "rev-list",
+                    f"--after={dandiset.version.created}",
+                    "-n1",
+                    "HEAD",
+                    cwd=repo,
+                )
+            ) != "":
```
Comment (yarikoptic) [inline tools/backups2datalad.py:494]: I would have preferred some more direct of "get direct children  of `committish` here.
also ain't with `--after={dandiset.version.created}` we might miss a commit which has exactly that stamp... I think it is unlikely to happen and likely the only commit with that date will be the tagged release commit (which will end up on the HEAD after #69 merged), which we should ignore here anyways.
So I guess we can proceed with this as is for now.

**Review by yarikoptic — CHANGES_REQUESTED** [review]
I ok'ed the logic on the commit consideration, but please encode at least some expectations from the fixed logic in the test:
- that release tags are produced off the main branch (not part of the HEAD)
- since you know what scenario it is, check that release commit is off the correct commit in the history
- their committed `dandiset.yaml` has doi filled in

**Comment by yarikoptic** [comment]
spotted that release commit is not having a correct date, I guess in all cases but spotted on 000056:
```
lena:/tmp
$> git -C 000056 show --format="%ad %cd %h %(describe)" -s master 0.210812.1518       
Thu Aug 12 11:28:25 2021 -0400 Thu Aug 12 11:28:25 2021 -0400 2e4c838 

tag 0.210812.1518
Tagger: DANDI Meta-user <dandi@mit.edu>

Version 0.210812.1518 of Dandiset 000056
Mon Aug 16 17:12:10 2021 -0400 Mon Aug 16 17:12:10 2021 -0400 e055b16 0.210812.1518
```

so it has current, not the one from Aug 12 whenever it was released.  Please 
- make sure that all `commit` and/or `ds.save` have the corresponding date changing context manager
- add testing of that into the test

File: tools/backups2datalad.py (hunk 1 above)
Comment (jwodder) [reply tools/backups2datalad.py:494]: Experimentation indicates that `--before` and `--after` will select a commit with exactly the given date & time.

**Comment by jwodder** [comment]
@yarikoptic Further tests added and tag date fixed.

**Comment by yarikoptic** [comment]
Didn't test in the "deployment" yet, but lets proceed and will test after all relevant PRs merged. Thank you @jwodder !

**Pull request merged**